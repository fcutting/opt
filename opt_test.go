package opt_test

import (
	"encoding/json"
	"fmt"
	"os"
	"testing"

	"github.com/fcutting/opt"
	"github.com/gkampitakis/go-snaps/snaps"
)

func TestMain(m *testing.M) {
	r := m.Run()
	snaps.Clean(m, snaps.CleanOpts{Sort: true})
	os.Exit(r)
}

type testPayload struct {
	Primitive opt.Option[string]         `json:"primitive"`
	Map       opt.Option[map[string]any] `json:"map"`
	Struct    opt.Option[testStruct]     `json:"struct"`
	Slice     opt.Option[[]int]          `json:"slice"`
	Pointer   opt.Option[*bool]          `json:"pointer"`
}

type testStruct struct {
	Make  string
	Model string
}

type testCase struct {
	data []byte
}

var (
	primitiveDefault = "default"
	mapDefault       = map[string]any{
		"make":  "Ford",
		"model": "Focus",
	}
	structDefault = testStruct{
		Make:  "Audi",
		Model: "A5",
	}
	sliceDefault   = []int{9, 8, 7}
	pointerDefault = true

	testCases = map[string]testCase{
		"Empty": {
			data: []byte(`
			{
				"empty": true
			}
		`),
		},
		"Primitive empty": {
			data: []byte(`
			{
				"primitive": ""
			}
		`),
		},
		"Primitive": {
			data: []byte(`
			{
				"primitive": "hello world"
			}
		`),
		},
		"Primitive null": {
			data: []byte(`
			{
				"primitive": null
			}
		`),
		},
		"Map empty": {
			data: []byte(`
			{
				"map": {}
			}
		`),
		},
		"Map full": {
			data: []byte(`
			{
				"map": {
					"make": "Toyota",
					"model": "Hilux"
				}
			}
		`),
		},
		"Map null": {
			data: []byte(`
			{
				"map": null
			}
		`),
		},
		"Struct empty": {
			data: []byte(`
			{
				"struct": {}
			}
		`),
		},
		"Struct full": {
			data: []byte(`
			{
				"struct": {
					"make": "Toyota",
					"model": "Hilux"
				}
			}
		`),
		},
		"Struct null": {
			data: []byte(`
			{
				"struct": null
			}
		`),
		},
		"Slice empty": {
			data: []byte(`
			{
				"slice": []
			}
		`),
		},
		"Slice full": {
			data: []byte(`
			{
				"slice": [1, 2, 3]
			}
		`),
		},
		"Slice null": {
			data: []byte(`
			{
				"slice": null
			}
		`),
		},
		"Pointer full": {
			data: []byte(`
			{
				"pointer": true
			}
		`),
		},
		"Pointer null": {
			data: []byte(`
			{
				"pointer": null
			}
		`),
		},
	}
)

func Test_Option(t *testing.T) {
	for n, c := range testCases {
		t.Run(n, func(t *testing.T) {
			var payload testPayload

			if err := json.Unmarshal(c.data, &payload); err != nil {
				t.Fatalf("Unexpected unmarshal error: %s", err)
			}

			t.Run("MarshalJSON", func(t *testing.T) {
				test_Marshal(t, payload)
			})

			t.Run("Exists", func(t *testing.T) {
				test_Exists(t, payload)
			})

			t.Run("Get", func(t *testing.T) {
				test_Get(t, payload)
			})

			t.Run("GetWithDefault", func(t *testing.T) {
				test_GetWithDefault(t, payload)
			})

			t.Run("String", func(t *testing.T) {
				test_String(t, payload)
			})
		})
	}
}

func test_Marshal(t *testing.T, payload testPayload) {
	result, err := json.Marshal(payload)

	if err != nil {
		t.Fatalf("Unexpected marshal error: %s", err)
	}

	snaps.MatchSnapshot(t, string(result))
}

func test_Exists(t *testing.T, payload testPayload) {
	t.Run("Primitive", func(t *testing.T) {
		snaps.MatchJSON(t, payload.Primitive.Exists())
	})

	t.Run("Map", func(t *testing.T) {
		snaps.MatchJSON(t, payload.Map.Exists())
	})

	t.Run("Struct", func(t *testing.T) {
		snaps.MatchJSON(t, payload.Struct.Exists())
	})

	t.Run("Slice", func(t *testing.T) {
		snaps.MatchJSON(t, payload.Slice.Exists())
	})

	t.Run("Pointer", func(t *testing.T) {
		snaps.MatchJSON(t, payload.Pointer.Exists())
	})
}

func test_Get(t *testing.T, payload testPayload) {
	t.Run("Primitive", func(t *testing.T) {
		snaps.MatchSnapshot(t, payload.Primitive.Get())
	})

	t.Run("Map", func(t *testing.T) {
		snaps.MatchJSON(t, payload.Map.Get())
	})

	t.Run("Struct", func(t *testing.T) {
		snaps.MatchJSON(t, payload.Struct.Get())
	})

	t.Run("Slice", func(t *testing.T) {
		snaps.MatchJSON(t, payload.Slice.Get())
	})

	t.Run("Pointer", func(t *testing.T) {
		snaps.MatchJSON(t, payload.Pointer.Get())
	})
}

func test_GetWithDefault(t *testing.T, payload testPayload) {
	t.Run("Primitive", func(t *testing.T) {
		snaps.MatchSnapshot(t, payload.Primitive.GetWithDefault(primitiveDefault))
	})

	t.Run("Map", func(t *testing.T) {
		snaps.MatchJSON(t, payload.Map.GetWithDefault(mapDefault))
	})

	t.Run("Struct", func(t *testing.T) {
		snaps.MatchJSON(t, payload.Struct.GetWithDefault(structDefault))
	})

	t.Run("Slice", func(t *testing.T) {
		snaps.MatchJSON(t, payload.Slice.GetWithDefault(sliceDefault))
	})

	t.Run("Pointer", func(t *testing.T) {
		snaps.MatchJSON(t, payload.Pointer.GetWithDefault(&pointerDefault))
	})
}

func test_String(t *testing.T, payload testPayload) {
	t.Run("Primitive", func(t *testing.T) {
		snaps.MatchSnapshot(t, fmt.Sprint(payload.Primitive))
	})

	t.Run("Map", func(t *testing.T) {
		snaps.MatchSnapshot(t, fmt.Sprint(payload.Map))
	})

	t.Run("Struct", func(t *testing.T) {
		snaps.MatchSnapshot(t, fmt.Sprint(payload.Struct))
	})

	t.Run("Slice", func(t *testing.T) {
		snaps.MatchSnapshot(t, fmt.Sprint(payload.Slice))
	})
}
