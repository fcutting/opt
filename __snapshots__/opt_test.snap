
[Test_Option/Empty/Exists/Map - 1]
false
---

[Test_Option/Empty/Exists/Primitive - 1]
false
---

[Test_Option/Empty/Exists/Slice - 1]
false
---

[Test_Option/Empty/Exists/Struct - 1]
false
---

[Test_Option/Empty/Get/Map - 1]
null
---

[Test_Option/Empty/Get/Primitive - 1]

---

[Test_Option/Empty/Get/Slice - 1]
null
---

[Test_Option/Empty/Get/Struct - 1]
{
 "Make": "",
 "Model": ""
}
---

[Test_Option/Empty/GetWithDefault/Map - 1]
{
 "make": "Ford",
 "model": "Focus"
}
---

[Test_Option/Empty/GetWithDefault/Primitive - 1]
default
---

[Test_Option/Empty/GetWithDefault/Slice - 1]
[
 9,
 8,
 7
]
---

[Test_Option/Empty/GetWithDefault/Struct - 1]
{
 "Make": "Audi",
 "Model": "A5"
}
---

[Test_Option/Empty/MarshalJSON - 1]
{"primitive":null,"map":null,"struct":null,"slice":null}
---

[Test_Option/Map_empty/Exists/Map - 1]
true
---

[Test_Option/Map_empty/Exists/Primitive - 1]
false
---

[Test_Option/Map_empty/Exists/Slice - 1]
false
---

[Test_Option/Map_empty/Exists/Struct - 1]
false
---

[Test_Option/Map_empty/Get/Map - 1]
{}
---

[Test_Option/Map_empty/Get/Primitive - 1]

---

[Test_Option/Map_empty/Get/Slice - 1]
null
---

[Test_Option/Map_empty/Get/Struct - 1]
{
 "Make": "",
 "Model": ""
}
---

[Test_Option/Map_empty/GetWithDefault/Map - 1]
{}
---

[Test_Option/Map_empty/GetWithDefault/Primitive - 1]
default
---

[Test_Option/Map_empty/GetWithDefault/Slice - 1]
[
 9,
 8,
 7
]
---

[Test_Option/Map_empty/GetWithDefault/Struct - 1]
{
 "Make": "Audi",
 "Model": "A5"
}
---

[Test_Option/Map_empty/MarshalJSON - 1]
{"primitive":null,"map":{},"struct":null,"slice":null}
---

[Test_Option/Map_full/Exists/Map - 1]
true
---

[Test_Option/Map_full/Exists/Primitive - 1]
false
---

[Test_Option/Map_full/Exists/Slice - 1]
false
---

[Test_Option/Map_full/Exists/Struct - 1]
false
---

[Test_Option/Map_full/Get/Map - 1]
{
 "make": "Toyota",
 "model": "Hilux"
}
---

[Test_Option/Map_full/Get/Primitive - 1]

---

[Test_Option/Map_full/Get/Slice - 1]
null
---

[Test_Option/Map_full/Get/Struct - 1]
{
 "Make": "",
 "Model": ""
}
---

[Test_Option/Map_full/GetWithDefault/Map - 1]
{
 "make": "Toyota",
 "model": "Hilux"
}
---

[Test_Option/Map_full/GetWithDefault/Primitive - 1]
default
---

[Test_Option/Map_full/GetWithDefault/Slice - 1]
[
 9,
 8,
 7
]
---

[Test_Option/Map_full/GetWithDefault/Struct - 1]
{
 "Make": "Audi",
 "Model": "A5"
}
---

[Test_Option/Map_full/MarshalJSON - 1]
{"primitive":null,"map":{"make":"Toyota","model":"Hilux"},"struct":null,"slice":null}
---

[Test_Option/Map_null/Exists/Map - 1]
true
---

[Test_Option/Map_null/Exists/Primitive - 1]
false
---

[Test_Option/Map_null/Exists/Slice - 1]
false
---

[Test_Option/Map_null/Exists/Struct - 1]
false
---

[Test_Option/Map_null/Get/Map - 1]
{}
---

[Test_Option/Map_null/Get/Primitive - 1]

---

[Test_Option/Map_null/Get/Slice - 1]
null
---

[Test_Option/Map_null/Get/Struct - 1]
{
 "Make": "",
 "Model": ""
}
---

[Test_Option/Map_null/GetWithDefault/Map - 1]
{}
---

[Test_Option/Map_null/GetWithDefault/Primitive - 1]
default
---

[Test_Option/Map_null/GetWithDefault/Slice - 1]
[
 9,
 8,
 7
]
---

[Test_Option/Map_null/GetWithDefault/Struct - 1]
{
 "Make": "Audi",
 "Model": "A5"
}
---

[Test_Option/Map_null/MarshalJSON - 1]
{"primitive":null,"map":{},"struct":null,"slice":null}
---

[Test_Option/Primitive/Exists/Map - 1]
false
---

[Test_Option/Primitive/Exists/Primitive - 1]
true
---

[Test_Option/Primitive/Exists/Slice - 1]
false
---

[Test_Option/Primitive/Exists/Struct - 1]
false
---

[Test_Option/Primitive/Get/Map - 1]
null
---

[Test_Option/Primitive/Get/Primitive - 1]
hello world
---

[Test_Option/Primitive/Get/Slice - 1]
null
---

[Test_Option/Primitive/Get/Struct - 1]
{
 "Make": "",
 "Model": ""
}
---

[Test_Option/Primitive/GetWithDefault/Map - 1]
{
 "make": "Ford",
 "model": "Focus"
}
---

[Test_Option/Primitive/GetWithDefault/Primitive - 1]
hello world
---

[Test_Option/Primitive/GetWithDefault/Slice - 1]
[
 9,
 8,
 7
]
---

[Test_Option/Primitive/GetWithDefault/Struct - 1]
{
 "Make": "Audi",
 "Model": "A5"
}
---

[Test_Option/Primitive/MarshalJSON - 1]
{"primitive":"hello world","map":null,"struct":null,"slice":null}
---

[Test_Option/Primitive_empty/Exists/Map - 1]
false
---

[Test_Option/Primitive_empty/Exists/Primitive - 1]
true
---

[Test_Option/Primitive_empty/Exists/Slice - 1]
false
---

[Test_Option/Primitive_empty/Exists/Struct - 1]
false
---

[Test_Option/Primitive_empty/Get/Map - 1]
null
---

[Test_Option/Primitive_empty/Get/Primitive - 1]

---

[Test_Option/Primitive_empty/Get/Slice - 1]
null
---

[Test_Option/Primitive_empty/Get/Struct - 1]
{
 "Make": "",
 "Model": ""
}
---

[Test_Option/Primitive_empty/GetWithDefault/Map - 1]
{
 "make": "Ford",
 "model": "Focus"
}
---

[Test_Option/Primitive_empty/GetWithDefault/Primitive - 1]

---

[Test_Option/Primitive_empty/GetWithDefault/Slice - 1]
[
 9,
 8,
 7
]
---

[Test_Option/Primitive_empty/GetWithDefault/Struct - 1]
{
 "Make": "Audi",
 "Model": "A5"
}
---

[Test_Option/Primitive_empty/MarshalJSON - 1]
{"primitive":"","map":null,"struct":null,"slice":null}
---

[Test_Option/Primitive_null/Exists/Map - 1]
false
---

[Test_Option/Primitive_null/Exists/Primitive - 1]
false
---

[Test_Option/Primitive_null/Exists/Slice - 1]
false
---

[Test_Option/Primitive_null/Exists/Struct - 1]
false
---

[Test_Option/Primitive_null/Get/Map - 1]
null
---

[Test_Option/Primitive_null/Get/Primitive - 1]

---

[Test_Option/Primitive_null/Get/Slice - 1]
null
---

[Test_Option/Primitive_null/Get/Struct - 1]
{
 "Make": "",
 "Model": ""
}
---

[Test_Option/Primitive_null/GetWithDefault/Map - 1]
{
 "make": "Ford",
 "model": "Focus"
}
---

[Test_Option/Primitive_null/GetWithDefault/Primitive - 1]
default
---

[Test_Option/Primitive_null/GetWithDefault/Slice - 1]
[
 9,
 8,
 7
]
---

[Test_Option/Primitive_null/GetWithDefault/Struct - 1]
{
 "Make": "Audi",
 "Model": "A5"
}
---

[Test_Option/Primitive_null/MarshalJSON - 1]
{"primitive":null,"map":null,"struct":null,"slice":null}
---

[Test_Option/Slice_empty/Exists/Map - 1]
false
---

[Test_Option/Slice_empty/Exists/Primitive - 1]
false
---

[Test_Option/Slice_empty/Exists/Slice - 1]
true
---

[Test_Option/Slice_empty/Exists/Struct - 1]
false
---

[Test_Option/Slice_empty/Get/Map - 1]
null
---

[Test_Option/Slice_empty/Get/Primitive - 1]

---

[Test_Option/Slice_empty/Get/Slice - 1]
[]
---

[Test_Option/Slice_empty/Get/Struct - 1]
{
 "Make": "",
 "Model": ""
}
---

[Test_Option/Slice_empty/GetWithDefault/Map - 1]
{
 "make": "Ford",
 "model": "Focus"
}
---

[Test_Option/Slice_empty/GetWithDefault/Primitive - 1]
default
---

[Test_Option/Slice_empty/GetWithDefault/Slice - 1]
[]
---

[Test_Option/Slice_empty/GetWithDefault/Struct - 1]
{
 "Make": "Audi",
 "Model": "A5"
}
---

[Test_Option/Slice_empty/MarshalJSON - 1]
{"primitive":null,"map":null,"struct":null,"slice":[]}
---

[Test_Option/Slice_full/Exists/Map - 1]
false
---

[Test_Option/Slice_full/Exists/Primitive - 1]
false
---

[Test_Option/Slice_full/Exists/Slice - 1]
true
---

[Test_Option/Slice_full/Exists/Struct - 1]
false
---

[Test_Option/Slice_full/Get/Map - 1]
null
---

[Test_Option/Slice_full/Get/Primitive - 1]

---

[Test_Option/Slice_full/Get/Slice - 1]
[
 1,
 2,
 3
]
---

[Test_Option/Slice_full/Get/Struct - 1]
{
 "Make": "",
 "Model": ""
}
---

[Test_Option/Slice_full/GetWithDefault/Map - 1]
{
 "make": "Ford",
 "model": "Focus"
}
---

[Test_Option/Slice_full/GetWithDefault/Primitive - 1]
default
---

[Test_Option/Slice_full/GetWithDefault/Slice - 1]
[
 1,
 2,
 3
]
---

[Test_Option/Slice_full/GetWithDefault/Struct - 1]
{
 "Make": "Audi",
 "Model": "A5"
}
---

[Test_Option/Slice_full/MarshalJSON - 1]
{"primitive":null,"map":null,"struct":null,"slice":[1,2,3]}
---

[Test_Option/Slice_null/Exists/Map - 1]
false
---

[Test_Option/Slice_null/Exists/Primitive - 1]
false
---

[Test_Option/Slice_null/Exists/Slice - 1]
false
---

[Test_Option/Slice_null/Exists/Struct - 1]
false
---

[Test_Option/Slice_null/Get/Map - 1]
null
---

[Test_Option/Slice_null/Get/Primitive - 1]

---

[Test_Option/Slice_null/Get/Slice - 1]
null
---

[Test_Option/Slice_null/Get/Struct - 1]
{
 "Make": "",
 "Model": ""
}
---

[Test_Option/Slice_null/GetWithDefault/Map - 1]
{
 "make": "Ford",
 "model": "Focus"
}
---

[Test_Option/Slice_null/GetWithDefault/Primitive - 1]
default
---

[Test_Option/Slice_null/GetWithDefault/Slice - 1]
[
 9,
 8,
 7
]
---

[Test_Option/Slice_null/GetWithDefault/Struct - 1]
{
 "Make": "Audi",
 "Model": "A5"
}
---

[Test_Option/Slice_null/MarshalJSON - 1]
{"primitive":null,"map":null,"struct":null,"slice":null}
---

[Test_Option/Struct_empty/Exists/Map - 1]
false
---

[Test_Option/Struct_empty/Exists/Primitive - 1]
false
---

[Test_Option/Struct_empty/Exists/Slice - 1]
false
---

[Test_Option/Struct_empty/Exists/Struct - 1]
true
---

[Test_Option/Struct_empty/Get/Map - 1]
null
---

[Test_Option/Struct_empty/Get/Primitive - 1]

---

[Test_Option/Struct_empty/Get/Slice - 1]
null
---

[Test_Option/Struct_empty/Get/Struct - 1]
{
 "Make": "",
 "Model": ""
}
---

[Test_Option/Struct_empty/GetWithDefault/Map - 1]
{
 "make": "Ford",
 "model": "Focus"
}
---

[Test_Option/Struct_empty/GetWithDefault/Primitive - 1]
default
---

[Test_Option/Struct_empty/GetWithDefault/Slice - 1]
[
 9,
 8,
 7
]
---

[Test_Option/Struct_empty/GetWithDefault/Struct - 1]
{
 "Make": "",
 "Model": ""
}
---

[Test_Option/Struct_empty/MarshalJSON - 1]
{"primitive":null,"map":null,"struct":{"Make":"","Model":""},"slice":null}
---

[Test_Option/Struct_full/Exists/Map - 1]
false
---

[Test_Option/Struct_full/Exists/Primitive - 1]
false
---

[Test_Option/Struct_full/Exists/Slice - 1]
false
---

[Test_Option/Struct_full/Exists/Struct - 1]
true
---

[Test_Option/Struct_full/Get/Map - 1]
null
---

[Test_Option/Struct_full/Get/Primitive - 1]

---

[Test_Option/Struct_full/Get/Slice - 1]
null
---

[Test_Option/Struct_full/Get/Struct - 1]
{
 "Make": "Toyota",
 "Model": "Hilux"
}
---

[Test_Option/Struct_full/GetWithDefault/Map - 1]
{
 "make": "Ford",
 "model": "Focus"
}
---

[Test_Option/Struct_full/GetWithDefault/Primitive - 1]
default
---

[Test_Option/Struct_full/GetWithDefault/Slice - 1]
[
 9,
 8,
 7
]
---

[Test_Option/Struct_full/GetWithDefault/Struct - 1]
{
 "Make": "Toyota",
 "Model": "Hilux"
}
---

[Test_Option/Struct_full/MarshalJSON - 1]
{"primitive":null,"map":null,"struct":{"Make":"Toyota","Model":"Hilux"},"slice":null}
---

[Test_Option/Struct_null/Exists/Map - 1]
false
---

[Test_Option/Struct_null/Exists/Primitive - 1]
false
---

[Test_Option/Struct_null/Exists/Slice - 1]
false
---

[Test_Option/Struct_null/Exists/Struct - 1]
false
---

[Test_Option/Struct_null/Get/Map - 1]
null
---

[Test_Option/Struct_null/Get/Primitive - 1]

---

[Test_Option/Struct_null/Get/Slice - 1]
null
---

[Test_Option/Struct_null/Get/Struct - 1]
{
 "Make": "",
 "Model": ""
}
---

[Test_Option/Struct_null/GetWithDefault/Map - 1]
{
 "make": "Ford",
 "model": "Focus"
}
---

[Test_Option/Struct_null/GetWithDefault/Primitive - 1]
default
---

[Test_Option/Struct_null/GetWithDefault/Slice - 1]
[
 9,
 8,
 7
]
---

[Test_Option/Struct_null/GetWithDefault/Struct - 1]
{
 "Make": "Audi",
 "Model": "A5"
}
---

[Test_Option/Struct_null/MarshalJSON - 1]
{"primitive":null,"map":null,"struct":null,"slice":null}
---
